### PR TITLE
set cfg(rustdoc) when rustdoc is running on a crate

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
When using `#[doc(cfg)]` to document platform-specific items, it's a little cumbersome to get all the platforms' items to appear all at once. For example, the standard library adds `--cfg dox` to rustdoc's command line whenever it builds docs, and the documentation for `#![feature(doc_cfg)]` suggests using a Cargo feature to approximate the same thing. This is a little awkward, because you always need to remember to set `--features dox` whenever you build documentation.

This PR proposes making rustdoc set `#[cfg(rustdoc)]` whenever it runs on a crate, to provide an officially-sanctioned version of this that is set automatically. This way, there's a standardized way to declare that a certain version of an item is specifically when building docs.

To try to prevent the spread of this feature from happening too quickly, this PR also restricts the use of this flag to whenever `#![feature(doc_cfg)]` is active. I'm sure there are other uses for this, but right now i'm tying it to this feature. (If it makes more sense to give this its own feature, i can easily do that.)

r? @rust-lang/rustdoc (and maaaaaaybe someone from core? This seems a bit far-reaching...)

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I'll bring this up at core team triage next week if I don't forget.

**Comment by GuillaumeGomez** [comment]
Awesome!

**Comment by nikomatsakis** [comment]
cc @rust-lang/lang -- this PR is adding a `#[cfg(rustdoc)]` so that people can know that rustdoc is running and tweak their crate as needed. Seems fine to me but though y'all might like to be aware.

**Comment by Mark-Simulacrum** [comment]
We discussed this in Core team triage and the conclusion was to cc the lang team; we felt it wasn't really our field of purview at this point.

**Comment by TimNN** [comment]
I've added `I-nominated` to make sure the @rust-lang/lang team takes a look.

**Comment by joshtriplett** [comment]
This makes me a *little* nervous, but given that people can trivially do this themselves anyway, we might as well provide a sanctioned solution.

That said...I wonder if rustdoc could somehow handle platform-specific things more automatically, rather than requiring people to use `cfg(rustdoc)`? rustdoc could learn to "see through" `cfg`, automatically document everything, and provide some readable description of the platforms an item is available on.

**Comment by QuietMisdreavus** [comment]
@joshtriplett You have just described [the oldest rustdoc issue on this tracker](https://github.com/rust-lang/rust/issues/1998).

The biggest problem with "making rustdoc see through cfg" is that *rustdoc* isn't the one that's processing them. *Rustc* is. Rustdoc will happily try to show off everything that's given to it, if rustc can handle ignoring every `#[cfg]` attribute in a crate and still make it through macro expansion and name resolution. Making that happen is something i have called ["the holy grail of rustdoc"](https://github.com/rust-lang/rust/issues/1998#issuecomment-318800909) because i'm pretty sure it will require way more work in the compiler then people are willing to deal with.

**Comment by joshtriplett** [comment]
@QuietMisdreavus Ah, I see. I didn't realize that, I thought this fell under the domain of rustdoc rather than rustc.

**Comment by Centril** [comment]
Seems fine to me; especially since it is feature gated. :)

**Comment by joshtriplett** [comment]
Chatted with @QuietMisdreavus about this at RustConf, and it seems reasonable. I'm going to go ahead and FCP this and see if anyone objects.

@rfcbot merge

**Comment by QuietMisdreavus** [comment]
@onur That should be fairly easy to set up, since you can add cfg flags via the command-line. We'd just need to add `--cfg docsrs` to the rustdoc flags whenever we call rustdoc over there and that should work out.

File: src/librustdoc/lib.rs [hunk 1]
```diff
@@ -645,7 +645,7 @@ where R: 'static + Send,
         paths.add_path(s, ErrorOutputType::default());
     }
     let mut cfgs = matches.opt_strs("cfg");
-    cfgs.push("rustdoc");
+    cfgs.push("rustdoc".to_string());
```
Comment (joshtriplett) [inline src/librustdoc/lib.rs]: Can you please squash this into the patch that introduced these lines?

Comment (QuietMisdreavus) [reply src/librustdoc/lib.rs]: Done!

**Comment by GuillaumeGomez** [comment]
Since everyone agreed, let's get it in then. Thanks @QuietMisdreavus!

@bors: r+

**Pull request merged**